quic/s2n-quic/src/stream/send.rs

```diff
@@ -577,6 +577,22 @@ impl SendStream {
         self.0.id().into()
     }
 
+    /// Set this stream to immediately terminate on any detected loss.
+    ///
+    /// This creates a stream that acts similar to an unreliable datagram, but is allowed to be
+    /// of arbitrary size. That means that s2n-quic will transparently handle the current MTU
+    /// to "fragment" the payload of this stream (like with any other stream), but if loss is
+    /// detected the stream is abandoned and will be immediately closed.
+    #[inline]
+    pub fn reset_on_loss(&mut self) {
```

Comment:
Toggling the value does not seem like a good idea to me. Supporting that to some extent implies knowing "how much" data was sent, for any useful disabling -- otherwise you don't really get reliability. And it means we can't easily do some of the optimizations suggested below, since e.g. even after FIN is sent you might disable the unreliability. So I don't think we should support that.

I like directly opening this as an unreliable stream, though it would be the same type presumably -- it would be a breaking change to add new [stream variants](https://docs.rs/s2n-quic/latest/s2n_quic/stream/enum.Stream.html). But that seems fine at least for now.

Yeah, I think making the name more generic is a good idea. In terms of QUIC extension -- for our use case, both ends know that all unidirectional streams are "unreliable", so maybe we can just signal that to both ends in our code? That would avoid network changes (which are trickier if we want to evolve specifics, though I'm not sure we would want to).

Reply:
> Toggling the value does not seem like a good idea to me.

I agree. Just wanted to make sure we at least documented why we're not going with that :)

> I like directly opening this as an unreliable stream, though it would be the same type presumably

Yeah we don't need a new type for it, since it's just a Send stream with some different internal logic.

> In terms of QUIC extension -- for our use case, both ends know that all unidirectional streams are "unreliable", so maybe we can just signal that to both ends in our code? That would avoid network changes (which are trickier if we want to evolve specifics, though I'm not sure we would want to).

I've been thinking a bit about this and we'd likely need 
* a new transport parameter to signal that each side understands the new unreliable behavior
* a new frame to create an unreliable stream:
    ```
    UNRELIABLE_STREAM Frame {
        Type (i) = 0xXXXX,
        Stream ID (i),
    }
    ```

    Receiving this frame would create and configure the stream as unreliable. Then the rest of the wire format would be like normal. One downside with this is not every packet would indicate that the stream is unreliable. For example, if the stream spanned 2 packets:

    ```
    Packet { number: 1, payload: [ UnreliableStream { id: 1 }, Stream { id: 1, offset: 0, len: 1200 } ] }
    Packet { number: 2, payload: [ Stream { id: 1, offset: 1200, len: 1000 } ] }
    ```

    If packet `1` was lost but received packet `2`, the receiver would create stream `1` and be waiting for the peer to process the ACK to retransmit, even though the stream was marked as unreliable. You could get around this by including the unreliable frame in front of every STREAM frame until you get an ACK or it transmit the FIN.

    ```
    Packet { number: 1, payload: [ UnreliableStream { id: 1 }, Stream { id: 1, offset: 0, len: 1200 } ] }
    Packet { number: 2, payload: [ UnreliableStream { id: 1 }, Stream { id: 1, offset: 1200, len: 1000 } ] }
    ```

    Another option would be creating a new frame that looks just like the [regular STREAM frame](https://www.rfc-editor.org/rfc/rfc9000#name-stream-frames), just with different IDs indicating it's unreliable.
    ```
    UNRELIABLE_STREAM Frame {
        Type (i) = 0xXXXX,
        Stream ID (i),
        [Offset (i)],
        [Length (i)],
        Stream Data (..),
    }
    ```

    The obvious downside is now there's 2 frames that essentially do the exact same thing, but every single one would carry all of the required information about its reliability.